src/qibolab/instruments/bluefors.py

```diff
@@ -0,0 +1,91 @@
+import json
+import re
+import socket
+from datetime import datetime
+
+from qibo.config import log
+
+from qibolab.instruments.abstract import Instrument
+
+
+class TemperatureController(Instrument):
+    """Bluefors temperature controller.
+
+    ```
+    # Example usage
+    if __name__ == "__main__":
+        tc = TemperatureController("XLD1000_Temperature_Controller", "192.168.0.114", 8888)
+        tc.connect()
+        temperature_values = tc.read_data()
+        for temperature_value in temperature_values:
+            print(temperature_value)
+    ```
+    """
+
+    def __init__(self, name: str, address: str, port: int = 8888):
+        """Creation of the controller object.
+
+        Args:
+            name (str): name of the instrument.
+            address (str): IP address of the board sending cryo temperature data.
+            port (int): port of the board sending cryo temperature data.
+        """
+        self.port = port
+        self.client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        super().__init__(name, address)
+
+    def connect(self):
+        """Connect to the socket."""
+        if self.is_connected:
+            return
+        log.info(f"Bluefors connection. IP: {self.address} Port: {self.port}")
+        self.client_socket.connect((self.address, self.port))
+        self.is_connected = True
+        log.info("Bluefors Temperature Controller Connected")
+
+    def disconnect(self):
+        """Disconnect from the socket."""
+        if self.is_connected:
+            self.client_socket.close()
+            self.is_connected = False
+
+    def setup(self):
+        """Required by parent class, but not used here."""
+        pass
+
+    @staticmethod
+    def convert_to_json(message: str) -> dict[str, dict[str, float]]:
+        """Convert the received socket message into a dictionary.
+
+        The typical message looks like this:
+            flange_name: {'temperature':12.345678, 'timestamp':1234567890.123456}
+        Args:
+            message (str): messaged received from the socket.
+        Returns:
+            dictionary_message (dict[str, dict[str, float]]):
+                message converted into python dictionary.
+        """
+        message = "\n".join(
+            [re.sub("^([^':]+)", r"'\g<1>'", m) for m in message.split("\n")]
+        )
+        message = re.sub("'", '"', message)
+        message = ",".join(message.split("\n"))
+        dictionary_message = json.loads("{" + message + "}")
+        for flange_values in dictionary_message.values():
+            flange_values["time"] = datetime.fromtimestamp(flange_values["timestamp"])
```

Comment:
Returning structured output is definitely a nice service to the user, but arguably you should not do it, because if this is the only way to retrieve those data, a user that doesn't care about the structuring will have to wait anyhow for it.

However, that's an abstract objection, in practice is good enough that is worth keeping it.
But I believe that's not the case for the `timestamp`: if you're asking "what is the temperature now?" it's likely that you won't even look at the time, it's just the most recent measurement you could get. I would avoid deserializing this, and leave it to the user.

```suggestion
```

Reply:
However, if you want to make a simpler service to the user, just document in the docstring that the `timestamp` is the type of `datetime.fromtimestamp` argument.

Reply:
Ok, I'll just leave the float value.